d2/src/main/java/com/linkedin/d2/balancer/clients/FailoutClient.java

```diff
@@ -0,0 +1,124 @@
+import com.linkedin.d2.balancer.D2ClientDelegator;
+import com.linkedin.d2.balancer.LoadBalancerWithFacilities;
+import com.linkedin.d2.balancer.ServiceUnavailableException;
+import com.linkedin.d2.balancer.clusterfailout.FailoutConfig;
+import com.linkedin.d2.balancer.util.LoadBalancerUtil;
+import com.linkedin.r2.message.Request;
+import com.linkedin.r2.message.RequestContext;
+import com.linkedin.r2.message.rest.RestRequest;
+import com.linkedin.r2.message.rest.RestResponse;
+import com.linkedin.r2.message.stream.StreamRequest;
+import com.linkedin.r2.message.stream.StreamResponse;
+import java.util.concurrent.Future;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A D2 delegator which rewrites URIs to redirect requests to another cluster if the target service has been failed out
+ * of service.
+ */
+public class FailoutClient extends D2ClientDelegator {
+  private static final Logger LOG = LoggerFactory.getLogger(FailoutClient.class);
+
+  private final FailoutRedirectStrategy _rewriteStrategy;
+  private final LoadBalancerWithFacilities _balancer;
+
+  public FailoutClient(D2Client d2Client, LoadBalancerWithFacilities balancer,
+      FailoutRedirectStrategy rewriteStrategy)
+  {
+    super(d2Client);
+    _balancer = balancer;
+    _rewriteStrategy = rewriteStrategy;
+  }
+
+  @Override
+  public Future<RestResponse> restRequest(final RestRequest request)
+  {
+    return restRequest(request, new RequestContext());
+  }
+
+  @Override
+  public Future<RestResponse> restRequest(final RestRequest request, final RequestContext requestContext)
+  {
+    final FutureCallback<RestResponse> future = new FutureCallback<>();
+    restRequest(request, requestContext, future);
+    return future;
+  }
+
+  @Override
+  public void restRequest(final RestRequest request, final Callback<RestResponse> callback)
+  {
+    restRequest(request, new RequestContext(), callback);
+  }
+
+  @Override
+  public void restRequest(RestRequest request, final RequestContext requestContext,
+      final Callback<RestResponse> callback)
+  {
+    final FailoutConfig config = getClusterFailoutConfig(request);
+    if (config != null && config.isFailedOut()) {
+      request = request.builder().setURI(
+          _rewriteStrategy.redirect(config.getPeerClusters(), request.getURI())).build();
+    }
+
+    super.restRequest(request, requestContext, callback);
+  }
+
+  @Override
+  public void streamRequest(final StreamRequest request, final Callback<StreamResponse> callback)
+  {
+    streamRequest(request, new RequestContext(), callback);
+  }
+
+  @Override
+  public void streamRequest(StreamRequest request, final RequestContext requestContext,
+      final Callback<StreamResponse> callback)
+  {
+    final FailoutConfig config = getClusterFailoutConfig(request);
+    if (config != null && config.isFailedOut()) {
+      request = request.builder().setURI(
+          _rewriteStrategy.redirect(config.getPeerClusters(), request.getURI())).build(
```

Comment:
i think we need the entire config here? We will match the failed out buckets against member's assigned bucket and then use redirect configs to figure out the peer cluster.

Reply:
I've made this change as it is a more generic interface and allows implementations to handle more cases. Is your expectation that we will do the bucket matching in the redirect, and if we don't see a the service's assigned bucket is in the failed out range the URI returned is the same as the original? In my current implementation, I am doing that logic in the `isFailedOut()` call, so the redirect call is only made if we have already determined a redirect is needed.

Reply:
I see. I'm thinking of `isFailedOut` as a short-circuiting check to see if the cluster has any active fail out vs. individual member. One example scenario where it would return false is when all the buckets are online but config has not been removed yet.